src/arena.c

```diff
@@ -883,6 +883,59 @@ arena_cache_bin_fill_small(tsdn_t *tsdn, arena_t *arena,
 	arena_decay_tick(tsdn, arena);
 }
 
+size_t
+arena_fill_small_fresh(tsdn_t *tsdn, arena_t *arena, szind_t binind,
+    void **ptrs, size_t nfill) {
+	if (nfill == 0) {
+		return 0;
+	}
+
+	assert(binind < SC_NBINS);
+	const bin_info_t *bin_info = &bin_infos[binind];
+	const size_t nfree = bin_info->nregs;
+	assert(nfree > 0);
+	unsigned binshard;
+	bin_t *bin = arena_bin_choose(tsdn, arena, binind, &binshard);
+	size_t nslab = 0;
+	size_t filled = 0;
+	edata_t *slab;
+
+	do {
```

Comment:
Ahhhh, I understand the logic now. I didn't see the only-the-last-iteration semantics. I think this code layout is fine in that case (but could you mind adding a comment explaining).

I could go either way on the optimization; I can make an argument I buy in both directions:
- Argument for doing it: this is the API people are calling because they're doing hot-path constant factor optimizations; we should try to make it as fast as we can.
- Argument against doing it: it's sort of a pain. This path is already faster than just calling malloc() N times, so it's not like anyone has a right to complain.

I think I lean towards doing the optimization, but I won't fight you on it if you think it's too much trouble for now.

I think the most reasonable place to get storage is probably to use an edata_list_active_t. There's an optimization within the optimization though -- we don't track full slabs for auto arenas (since auto arenas can't be destroyed, which is the only time we need to track them). So we can even avoid iterating through allocated slabs in the common cases (though we still need to track non-full slabs and stick them into slabs-nonfull). In the non-auto-arena case, we can just do a batch list merge of the new full slabs into the arena bin slabs_full list, so in either case we only really need to touch a few unusual slabs.